### PR TITLE
feat(gil): Implement Tasks Feature

### PR DESCRIPTION
**Pull request by Skillz4Killz**
Please describe the changes this PR makes and why it should be merged:

- Schedule tasks to run at certain times. Very basic needs for 99% of bots. If something more is needed for the rare edge case user, just customize and setup as needed.

Status

-   [ ] Code changes have been tested.

Semantic versioning classification:

-   [ ] This PR changes the library's interface (methods or parameters added)
-   [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
-   [ ] This PR only includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: packages/gil/lib/BotClient.ts [hunk 1]
```diff
@@ -121,19 +127,61 @@ export class BotClient extends Client {
...
+    /** Allows users to override and customize the initialization of scheduled task intervals. */
+    initializeTasks(): void {
+        this.tasks.forEach(async (task) => {
+            // TASKS THAT NEED TO RUN IMMEDIATELY ARE EXECUTED FIRST
+            if (task.runOnStartup) await task.execute();
+
+            // SET TIMEOUT WILL DETERMINE THE RIGHT TIME TO RUN THE TASK FIRST TIME AFTER STARTUP
+            setTimeout(async () => {
+                console.log(`${bgBlue(`[${this.getTime()}]`)} [TASK: ${bgYellow(black(task.name))}] started.`);
+                try {
```
Comment (DaStormer) [inline packages/gil/lib/BotClient.ts]: There's no check for requiring ready here

File: packages/gil/lib/structures/Task.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import type { BotClient } from "../BotClient";
+
+export abstract class Task {
+    /** The amount of time this task should take to run. Defaults to 1 hour(3,600,000 ms) */
+    millisecondsInterval = 60 * 60 * 100;
```
Comment (DaStormer) [inline packages/gil/lib/structures/Task.ts]:
```suggestion
    millisecondsInterval = 60 * 60 * 1000;
```

File: packages/gil/lib/structures/Task.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    /** Whether or not this task should run immediately on startup. Default to false. */
+    runOnStartup = false;
+    /** Whether this task requires the bot to be fully ready before running. Default to false. */
+    requireFullyReady = false;
```
Comment (DaStormer) [inline packages/gil/lib/structures/Task.ts]: I'm being a little picky here, but I feel like `requireFullyReady` is too verbose, can be simplified to `requireReady`

**Review by DaStormer — CHANGES_REQUESTED** [review]
Awesome feature! Just a few things I noticed

File: packages/gil/lib/BotClient.ts [hunk 4]
```diff
@@ -139,44 +139,38 @@ export class BotClient extends Client {
     initializeTasks(): void {
         this.tasks.forEach(async (task) => {
             // TASKS THAT NEED TO RUN IMMEDIATELY ARE EXECUTED FIRST
-            if (task.runOnStartup) await task.execute();
+            if (task.runOnStartup) await this.executeTask(task);
 
             // SET TIMEOUT WILL DETERMINE THE RIGHT TIME TO RUN THE TASK FIRST TIME AFTER STARTUP
             setTimeout(async () => {
```
Comment (zaida04) [inline packages/gil/lib/BotClient.ts:145]: Could this not just be condensed into the setInterval call after the initial call in the lines above this?

**Review by zaida04 — CHANGES_REQUESTED** [review]

**Review by DaStormer — APPROVED** [review]
other than nico's request, LGTM

**Review by zaida04 — APPROVED** [review]

**Pull request merged**